src/Microsoft.Azure.SignalR.Management/NegotiateProcessor.cs

```diff
@@ -11,9 +13,44 @@ namespace Microsoft.Azure.SignalR.Management
 {
     internal class NegotiateProcessor
     {
-        public Task<NegotiationResponse> GetClientEndpointAsync(string hubName, HttpContext httpContext = null, string userId = null, CancellationToken cancellationToken = default)
+        private readonly IServiceEndpointManager _serviceEndpointManager;
+        private readonly IEndpointRouter _router;
+        private readonly MultiEndpointConnectionContainerFactory _endpointsContainerFactory;
+
+        public NegotiateProcessor(IServiceEndpointManager serviceEndpointManager, IEndpointRouter router, MultiEndpointConnectionContainerFactory endpointsContainerFactory)
+        {
+            _serviceEndpointManager = serviceEndpointManager;
+            _router = router;
+            _endpointsContainerFactory = endpointsContainerFactory;
+        }
+
+        public async Task<NegotiationResponse> GetClientEndpointAsync(string hubName, HttpContext httpContext = null, string userId = null, IEnumerable<Claim> claims = null, TimeSpan? lifeTime = null, CancellationToken cancellationToken = default)
         {
-            throw new NotImplementedException();
+            if (cancellationToken == default && httpContext != null)
+            {
+                cancellationToken = httpContext.RequestAborted;
+            }
+
+            var container = _endpointsContainerFactory.GetOrCreate(hubName);
+            _ = container.StartAsync();
+            //ensure connections to each endpoint are initialized, so that the online status of endpoints are valid
+            await container.ConnectionInitializedTask.OrTimeout(cancellationToken);
+
+            var candidateEndpoints = _serviceEndpointManager.GetEndpoints(hubName);
+            var selectedEndpoint = _router.GetNegotiateEndpoint(httpContext, candidateEndpoints);
+            var provider = _serviceEndpointManager.GetEndpointProvider(selectedEndpoint);
+
+            userId ??= httpContext?.User?.Identity?.Name;
+            claims ??= httpContext?.User?.Claims;
+            var claimsWithUserId = ClaimsUtility.CreateUserClaims(userId, claims);
+
+            var tokenTask = provider.GenerateClientAccessTokenAsync(hubName, claimsWithUserId, lifeTime);
+            await tokenTask.OrTimeout(cancellationToken);
+            return new NegotiationResponse
+            {
+                Url = provider.GetClientEndpoint(hubName, null, null),
+                AccessToken = tokenTask.Result
```

Comment:
what happens when tokenTask is canceled? does it still come to this line trying to get the result?

Reply:
now an error is thrown when task is canceled.